src/model/nets/models_genesis.py

```diff
@@ -38,20 +40,43 @@ def __init__(self, in_channels, out_channels, weight_path=None, with_decoder=Tru
             state_dict = self.state_dict()
             if Path(weight_path).name == 'models_genesis.pth':
                 pretrained_state_dict = torch.load(weight_path, map_location='cpu')
-            else:  # For our pre-trained weight and the adapter fine-tine experiments.
+            else:  # For our saved model checkpoints.
                 pretrained_state_dict = torch.load(weight_path, map_location='cpu')['net']
                 pretrained_state_dict.pop('out_block.weight')
                 pretrained_state_dict.pop('out_block.bias')
-            if not with_decoder:
+            if loaded_modules is not None:
+                assert len(loaded_modules) > 0
+                loaded_modules = set(loaded_modules)
+                if 'encoder' in loaded_modules:
+                    loaded_modules.remove('encoder')
+                    loaded_modules.update({'in_block', 'down_block1', 'down_block2', 'down_block3'})
+                if 'decoder' in loaded_modules:
+                    loaded_modules.remove('decoder')
+                    loaded_modules.update({'up_block1', 'up_block2', 'up_block3'})
+                all_modules = {'in_block', 'down_block1', 'down_block2', 'down_block3',
+                               'up_block1', 'up_block2', 'up_block3'}
+                assert loaded_modules.issubset(all_modules)
                 for key in list(pretrained_state_dict.keys()):
-                    if 'up_block' in key:
+                    if not any(loaded_module in key for loaded_module in loaded_modules):
                         pretrained_state_dict.pop(key)
-
-            pretrained_state_dict = {key: value for key, value in pretrained_state_dict.items()
-                                     if key in state_dict.keys()}
             state_dict.update(pretrained_state_dict)
             self.load_state_dict(state_dict)
 
+        if frozen_modules is not None:
+            assert 'out_block' not in frozen_modules
+            frozen_modules = set(frozen_modules)
+            if 'encoder' in frozen_modules:
+                frozen_modules.remove('encoder')
+                frozen_modules.update(['in_block', 'down_block1', 'down_block2', 'down_block3'])
+            if 'decoder' in frozen_modules:
+                frozen_modules.remove('decoder')
+                frozen_modules.update(['up_block1', 'up_block2', 'up_block3'])
+            for param in itertools.chain.from_iterable(
```

Comment:
Maybe we can check whether the whole `frozen_modules` are valid instead of checking `out_block` only.
```
assert frozen_modules.issubset(all_modules)
```

Reply:
Good points, I will fix it.

Reply:
Fixed in 9d0a376.